feat : 리펙토링

cloudwi opened the pull request:
- record를 활용한 DTO 객체 리펙토링
- Member도메인 설계 시작
- TelegramBot 도메인 불필요객체 제거

cloudwi committed:
feat : 리펙토링

- record를 활용한 DTO 객체 리펙토링
- Member도메인 설계 시작
- TelegramBot 도메인 불필요객체 제거

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/controller/ProductController.java:
```diff
@@ -31,7 +32,10 @@ public ProductController(ProductService productService) {
     public ResponseEntity<ProductSaveResponseDto> save(
-        @RequestBody @Validated ProductSaveRequestDto productSaveRequestDto) {
+        @RequestBody
+        @Validated
```
매개변수 유효성 검사에 `@Valid` 가 아닌 `@Validated`를 사용하신 이유가 있을까요??

제가 알기론, `@Valid`는 자바 표준으로 지금 상황처럼 요청의 body를 검증하는데 많이 사용되고,
`@Validated`는 AOP 기반으로 그룹 유효성 검증을 위해 클래스 선언부에 사용하는 것으로 알고 있습니다.

좋은 포스팅이 있어 첨부합니다.
- https://mangkyu.tistory.com/174
- https://medium.com/sjk5766/valid-vs-validated-%EC%A0%95%EB%A6%AC-5665043cd64b

cloudwi replied:
덕분에 좋은거 알아가는군요 ㅎㅎ 
향후 그룹화가 필요하다면 고려해보겠습니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/service/impl/ProductTelegramServiceImpl.java:
```diff
@@ -1,49 +1,56 @@
+
     private final ProductRepository productRepository;
-    private Document doc;
+    private static final String BASE_UTL = "https://www.musinsa.com/app/goods/";
```
URL을 작성하려고 하신 것 같은데, 오타가 있는 것 같습니다.

cloudwi replied:
오타 발견 감사합니다.

2dongyeop reviewed (commented):
추가로 Request와 Response가 Dto 의미에 포함된다고 볼 수 있을 것 같은데,
Request 나 Dto가 아닌 RequestDto라는 이름은 중복된 표현이 아닐지 고민이 되네요. 
저도 찾아보도록 하겠습니다. 고생하셨어요!

cloudwi committed:
feat : 리펙토링

- 회원도메인 설계 시작하기
- ProductId -> ProductInfo
- @Validated 목적에 부합하지 않아서 -> @Valid
- 정적 팩토리 메서드 도입
- enum을 도입하여 TelegramBotCommand 명령어의 식별성을 높임
- 변수, 메서드 명을 직관적으로 수정
- 상수를 정의하여 하드코딩하지 않고 관리

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/telegramBot/service/TelegramService.java:
```diff
@@ -113,17 +118,23 @@ public void onUpdateReceived(Update update) {
         }
     }
 
+    private static ProductInfo getProductInfo(String[] inputMessage) {
```
해당 메소드가 공통 메서드로 분리하여 중복 코드를 줄인 것으로 보이는데 혹시 `static`으로 할 이유가 있을까요??

cloudwi replied:
TelegramService 안에서만 사용하니 메모리 공간 낭비를 방지하고자 static을 제거하는것이 좋을 것 같군요 감사합니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/telegramBot/service/TelegramService.java:
```diff
@@ -58,31 +59,34 @@ public String getBotToken() {
+        switch (telegramBotCommand) {
+            case ADD -> {
+                ProductInfo productInfo = getProductInfo(inputMessage);
+                long productId = productInfo.getId();
```
해당 메소드 이름이 `onUpdateReceived()` 이므로, 요청을 판별하는 핸들러 역할로 코드를 작성하신 것 같은데
모든 경우의 처리 로직을 이 메서드에서 다루다보니, `단일 책임 원칙`을 위반하는 것 같습니다.

해당 메서드에서는 핸들러로써 작업 판독을 하고, 처리 로직은 별도로 분리하는 것은 어떨까요?

cloudwi replied:
감사합니다. 

다른 service레이어의 책임까지 함께 가지고 있었던 것을 분리 하였습니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/service/impl/ProductTelegramServiceImpl.java:
```diff
@@ -18,28 +18,31 @@
 
     private final ProductRepository productRepository;
-    private static final String BASE_UTL = "https://www.musinsa.com/app/goods/";
+    private static final String BASE_URL = "https://www.musinsa.com/app/goods/";
+
+    private static final String TITLE_HTML = "span.product_title";
+    private static final String IMAGE_URL_LINK = "div.product-img";
```
해당 상수들이 ServiceImpl 클래스 내에서만 사용되고, `static` 메서드에서 사용하는 경우도  `static`일 필요는 없을 것 같습니다.
(변동사항 코드만 보고 리뷰를 적기에 다른곳에서도 사용하는 코드라면 무시하셔도 됩니다!)

cloudwi replied:
감사합니다.

특정 플랫폼(무신사)에 종속적인 코드입니다. 
따라서 지역변수로 사용하도록하겠습니다. 리펙토링 완료했습니다.

2dongyeop reviewed (commented):
늦은 시간에 고생하셨습니다~

ybkim-dev commented on src/main/java/com/project/musinsastocknotificationbot/product/controller/ProductController.java:
```diff
@@ -30,33 +32,41 @@ public ProductController(ProductService productService) {
-    public ResponseEntity<List<ProductFindAllResponseDto>> findAll() {
-        List<ProductFindAllResponseDto> productFindAllResponseDtos = new ArrayList<>();
+    public ResponseEntity<List<ProductFindAllResponse>> findAll() {
+        List<ProductFindAllResponse> productFindAllResponses = new ArrayList<>();
```
Product를 전체 조회할 때와 개별 조회할 때의 DTO가 다른가요?
다르지 않다면 어차피 List로 관리되고 있는데 하나의 DTO로 통일하면 어떨까요?

cloudwi replied:
감사합니다.

현재 서비스에 컨트롤러가 필요없다고 판단해서 DTO자체를 삭제하였습니다.

ybkim-dev commented on src/main/java/com/project/musinsastocknotificationbot/product/domain/dto/Response/ProductFindAllResponse.java:
```diff
@@ -0,0 +1,16 @@
+
+import com.project.musinsastocknotificationbot.product.domain.Product;
+
+public record ProductFindAllResponse(
+    long id,
+    String size,
+    String title,
+    String imageHTML
+) {
```
마찬가지로 네이밍을 변경해주셔도 될 것 같습니다.

cloudwi replied:
감사합니다. 😀

DTO 자체를 삭제하여서 다음 리펙토링 혹은 개발때 반영하겠습니다.

ybkim-dev commented on src/main/java/com/project/musinsastocknotificationbot/product/error/JsoupIOException.java:
```diff
@@ -0,0 +1,7 @@
+package com.project.musinsastocknotificationbot.product.error;
+
+public class JsoupIOException extends RuntimeException {
+
+  public JsoupIOException(Exception e) {
+  }
+}
```
해당 예외가 어떤 내용에 의해 발생하는 예외인지 파악할 수 있게 커스텀하셔도 될 것 같아요.

cloudwi replied:
커스텀은 향후에 rest api로 업데이트할때 반영하겠습니다. 감사합니다.

ybkim-dev commented on src/main/java/com/project/musinsastocknotificationbot/product/service/ProductService.java:
```diff
@@ -1,15 +1,15 @@
 public interface ProductService {
 
-    ProductId save(long id, String size);
+    ProductInfo save(long id, String size);
```
전에 이야기했듯이 id, size 외의 다른 컬럼이 추가되는 경우에 대해서는 문제가 생길 것 같아요. 뭉탱이로 만들어 파라미터로 받으면 어떨까요?

cloudwi replied:
감사합니다. 랩핑해서 사용하도록 하겠습니다 굿 입니다.

ybkim-dev commented on src/main/java/com/project/musinsastocknotificationbot/product/service/impl/ProductTelegramServiceImpl.java:
```diff
@@ -1,49 +1,60 @@
+    private static final String BASE_URL = "https://www.musinsa.com/app/goods/";
+
+    private static final String TITLE_HTML = "span.product_title";
+    private static final String IMAGE_URL_LINK = "div.product-img";
```
👍

cloudwi committed:
refactoring : BaseUrl enum 생성

- 서비스가 확장될 가능성을 생각하며 BaseUrl 상수들을 enum으로 관리하도록 설계

cloudwi committed:
refactoring : Crawling 객체 분리

- 서비스의 확장성을 고려하여 Crawling 객체를 분리하여 책임을 분산 시켰습니다.
- Crawling 인터페이스를 두어서 다양한 사이트의 제품을 크롤링할 수 있을 것 입니다.

cloudwi committed:
refactoring : Crawling 객체 분리

- 서비스의 확장성을 고려하여 Crawling 객체를 분리하여 책임을 분산 시켰습니다.
- Crawling 인터페이스를 두어서 다양한 사이트의 제품을 크롤링할 수 있을 것 입니다.

cloudwi committed:
refactoring : ProductService 무신사 구현체 제작

- 무신사이외의 사이트에 서비스하기위한 발판을 마련했습니다.

cloudwi committed:
refactoring : ProductService 무신사 구현체 제작

- 무신사이외의 사이트에 서비스하기위한 발판을 마련했습니다.

cloudwi committed:
refactoring : ProductInfo의 ID를 사용하는 Url을 반환하는 책임을 ProductInfo안으로 넣었습니다.

cloudwi committed:
refactoring : TelegramService의 책임을 다른 Service레이어로 전달했습니다.

cloudwi committed:
refactoring : TelegramService SendMessage의 중복이 강하여 정적변수로 등록했습니다.

bigtyno931128 commented on src/main/java/com/project/musinsastocknotificationbot/product/domain/vo/ProductInfo.java:
안녕하세요 ~! 
@Embeddable 사용 이유와 직렬화 사용 이유가 궁금합니다 !

cloudwi replied:
제품의 키값을 고정할때 제품의 키랑 사이즈가 일치해야합니다. 그래서 사이즈별로 알림이 발송가능합니다.

bigtyno931128 reviewed (commented):
코드가 굉장히 통일성 있게 보여서 좋았던것 같습니다 .

cloudwi committed:
feat : profiles 추가

- email 기능 추가 예정 작업

cloudwi committed:
refactor : Crawling 기능 분리

- 크롤링 기능을 분리해서 책임을 분산 시킴

cloudwi committed:
refactor : Crawling 기능 분리

- 크롤링 기능을 분리해서 책임을 분산 시킴

cloudwi committed:
refactor : Crawling 기능 분리

- 크롤링 기능을 분리해서 책임을 분산 시킴

cloudwi committed:
feat : Email 기능 서비스 구현

- 네이버 stmp 사용 예정

cloudwi committed:
refactor : 전체적인 레거시 코드를 리팩토링함

- 객체지향 원칙 준수를 목적으로 리펙토링 진행
- 새로운 기능 추가로 인한 영향력을 줄이기 위해 리펙토링 진행
- 버전 2로 업데이트 예정

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/infrastructure/impl/MusinsaCrawlingImpl.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class MusinsaCrawlingImpl {
```
Impl 클래스에 Crawling 인터페이스 구현(implements)이 빠진건 아닐까요?

2dongyeop replied:
아래 커밋을 보니 수정하셨네요! 무시하셔도 됩니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/infrastructure/impl/MusinsaCrawlingImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+      Document document = Jsoup.connect(url).get();
+      String title = document.select(titleHTML).text();
+      String imageUrl = document.select(imageUrlLink).html();
```
해당 로직이 저장과 관련된 로직이 아닌데 `@Repository` 애너테이션을 사용한 이유가 있을까요?
외부로부터 읽어오는 작업이기 때문에, 컨트롤러에 작성하는게 맞지 않을까 싶습니다.

+이 리뷰도 아래 커밋을 보니 수정하셨네요 무시하셔도 됩니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/product/service/impl/MusinsaProductServiceImpl.java:
```diff
@@ -45,7 +43,6 @@ public void findAll() {
         List<Product> products = productRepository.findAll();
 
         StringBuilder stringBuilder = new StringBuilder();
```
현재 메시지를 보내는 작업이 단일 스레드로 동작하고 있다면, StringBuilder를 사용하는 것이 성능이 더 좋겠지만
추후에 이 작업을 멀티스레드로 동작하도록 하신다면 StringBuffer 사용으로 고려해야 할 포인트인 것 같습니다.!

cloudwi replied:
감사합니다 버퍼로 동작하게 수정하였습니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/message/infrastructure/email/service/EmailService.java:
```diff
@@ -0,0 +1,27 @@
+    this.from = from;
+  }
+
+  @Override
+  public void sendMessage(String message) {
+
+  }
+}
```
추후에 작업시 `@Transactional` 애너테이션 추가가 필요해보입니다.
이메일 송신 작업의 경우는 기본적으로 다른 기능보다 작업 소요 시간이 오래되어, 멀티스레드 환경에서의 비동기 성능 개선이 반드시 필요합니다.

해당 기능에서 고민한 과정을 블로그에 포스팅한 경험이 있어, 부족하지만 첨부해봅니다.
- [velog : CompletableFuture로 비동기 작업 안정화 및 성능 테스트](https://velog.io/@dongvelop/CompletableFuture%EB%A1%9C-%EB%B9%84%EB%8F%99%EA%B8%B0-%EC%9E%91%EC%97%85-%EC%95%88%EC%A0%95%ED%99%94-%EB%B0%8F-%EC%84%B1%EB%8A%A5-%ED%85%8C%EC%8A%A4%ED%8A%B8)

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/message/infrastructure/telegramBot/service/TelegramMessageServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+        this.chatId = chatId;
+    }
+
+    public void sendMessage(String message) {
+        SendMessage sendMessage = new SendMessage();
+        sendMessage.setChatId(chatId);
+        sendMessage.setText(message);
+
+        try {
+            execute(sendMessage);
+        } catch (TelegramApiException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
SendMessage 객체를 만드는 과정을 `createMessage()` 등과 같이 메소드로 분리할 수 있을 것 같습니다.
현재로는 메시지를 만드는 로직이 변경될 경우에 메시지를 보내는 로직도 영향을 받게 되어 `단일 책임 원칙`에 위반되는 것 같아요.

cloudwi replied:
감사합니다. 메서드로 분리해서 향후 재 사용성과 책임을 분리 하였습니다.

2dongyeop commented on src/main/java/com/project/musinsastocknotificationbot/message/infrastructure/telegramBot/webClient/TelegramWebClient.java:
```diff
@@ -0,0 +1,28 @@
+public class TelegramWebClient {
+
+  private final WebClient webClient;
+  DefaultUriBuilderFactory defaultUriBuilderFactory;
```
`private` 접근 제한자를 빠트리신건 아닐지 생각이 듭니다!

cloudwi replied:
webClient를 사용하지 않더라구요 ㅎㅎ 삭제하였습니다. 감사합니다.

cloudwi committed:
refactor : 전체적인 코드를 리펙토링하였습니다.

- 기존 레거시 코드를 객체지향원칙을 준수하며 리펙토링
- 책임이 한 곳에 몰려있는 것을 분산 시킴
- 각 기능별 패키지 구조 개선

cloudwi committed:
refactor : 전체적인 코드를 리펙토링하였습니다.

- 기존 레거시 코드를 객체지향원칙을 준수하며 리펙토링
- 책임이 한 곳에 몰려있는 것을 분산 시킴
- 각 기능별 패키지 구조 개선
